### PR TITLE
:sparkles: supporting cloudevents for work agent

### PR DESCRIPTION
**Pull request by skeeey**
## Related issue(s)

Fixes #

### CODE REVIEW

File: pkg/work/spoke/spokeagent.go [hunk 1]
```diff
@@ -145,49 +139,101 @@ func (o *WorkAgentConfig) RunWorkloadAgent(ctx context.Context, controllerContex
...
 	go appliedManifestWorkController.Run(ctx, 1)
 	go manifestWorkController.Run(ctx, 1)
 	go manifestWorkFinalizeController.Run(ctx, manifestWorkFinalizeControllerWorkers)
 	go availableStatusController.Run(ctx, availableStatusControllerWorkers)
+
 	<-ctx.Done()
+
 	return nil
 }
+
```
Comment (qiujian16) [inline pkg/work/spoke/spokeagent.go:191]: it needs comment here. we should explain why we build like this.

Comment (skeeey) [reply pkg/work/spoke/spokeagent.go:191]: added

File: pkg/work/spoke/options.go [hunk 2]
```diff
@@ -4,19 +4,22 @@ import (
...
 
 	"github.com/spf13/pflag"
+	"open-cluster-management.io/api/cloudevents/generic/options/mqtt"
 )
 
 // WorkloadAgentOptions defines the flags for workload agent
 type WorkloadAgentOptions struct {
 	StatusSyncInterval                     time.Duration
 	AppliedManifestWorkEvictionGracePeriod time.Duration
+	MQTTOptions                            *mqtt.MQTTOptions
```
Comment (clyang82) [inline pkg/work/spoke/options.go:14]: Do we need to start webhook server if it is configured to run with mqtt option? I think we may not. https://github.com/open-cluster-management-io/ocm/blob/8a4c834ebf0f3c71ab9209572fb0b11cc942f5d5/cmd/work/main.go#L62

**Comment by clyang82** [comment]
How to test the MQTT protocol here?

File: pkg/work/spoke/options.go (hunk 2 above)
Comment (skeeey) [reply pkg/work/spoke/options.go:14]: this is options only for agent，the webhook-server is a sub command for work agent, we can control whether to start it outside

```
./work --help
Spoke Cluster Work

Available Commands:
  agent          Start the Work Agent
  completion     Generate the autocompletion script for the specified shell
  help           Help about any command
  manager        Start the Work Hub Manager
  webhook-server Start the work webhook server
```

to use this `work agent --mqtt-broker=...`

**Comment by skeeey** [comment]
> How to test the MQTT protocol here?

I'm not sure, the work client with MQTT is already tested in the api repo, but for work agent, now we don't have a server supporting this, currently, we may not be able to start the e2e test in the ocm, after the maestro server is ready, we may use it to do the e2e test in the ocm?

**Review by morvencao — COMMENTED** [review]
Do we also need to add cloudevents support to the singleton agent ? https://github.com/morvencao/OCM/tree/main/pkg/singleton/spoke

**Comment by skeeey** [comment]
> Do we also need to add cloudevents support to the singleton agent ? https://github.com/morvencao/OCM/tree/main/pkg/singleton/spoke

the singleton agent calls the work agent code, so this support will be also in the singleton agent

**Comment by qiujian16** [comment]
TODO, instead of using flags. It will be better to have a config file for mqtt setup. We should only have one flag to specify driver for the work agent

**Comment by skeeey** [comment]
> TODO, instead of using flags. It will be better to have a config file for mqtt setup. We should only have one flag to specify driver for the work agent

good suggestion, add the TODO to the code firstly :)

**Pull request merged**